File: app/src/main/java/top/asdb/codexremote/agent/OpenCodeAgentClient.kt
package top.asdb.codexremote.agent

import java.util.concurrent.ConcurrentHashMap
import kotlinx.coroutines.CoroutineScope
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.add
import kotlinx.serialization.json.booleanOrNull
import kotlinx.serialization.json.buildJsonArray
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put
import top.asdb.codexremote.BuildConfig
import top.asdb.codexremote.codex.CodexAppServerClient
import top.asdb.codexremote.codex.string
import top.asdb.codexremote.data.AgentConnectionTestResult
import top.asdb.codexremote.data.AgentCapabilities
import top.asdb.codexremote.data.AgentGlobalSettings
import top.asdb.codexremote.data.AgentKind
import top.asdb.codexremote.data.AgentMode
import top.asdb.codexremote.data.ApiModelOption
import top.asdb.codexremote.data.CustomModelDefinition
import top.asdb.codexremote.data.ServerProfile
import top.asdb.codexremote.data.modelSettings
import top.asdb.codexremote.ssh.RemoteBootstrap
import top.asdb.codexremote.ssh.RemoteCodexSettings
import top.asdb.codexremote.ssh.RemoteInstallProgress
import top.asdb.codexremote.ssh.SshCodexTransport

internal const val OPENCODE_MANAGED_PROVIDER_ID = "custom-api"
internal const val OPENCODE_LEGACY_MANAGED_PROVIDER_ID = "codex-remote"
internal val OPENCODE_REASONING_EFFORTS = listOf("minimal", "low", "medium", "high", "xhigh")
private val OPENCODE_REASONING_MODEL_PATTERN =
    Regex("(^|/)gpt-5(?:[._-]|$)|(^|/)o(?:1|3|4)(?:[._-]|$)", RegexOption.IGNORE_CASE)

internal fun openCodeReasoningEfforts(modelId: String): List<String> =
    OPENCODE_REASONING_EFFORTS.takeIf { OPENCODE_REASONING_MODEL_PATTERN.containsMatchIn(modelId.trim()) }
        .orEmpty()

internal fun normalizeOpenCodeModelId(value: String): String {
    val normalized = RemoteCodexSettings.normalizeDefaultModel(value)
    return when {
        normalized.isBlank() -> normalized
        normalized.startsWith("$OPENCODE_LEGACY_MANAGED_PROVIDER_ID/") ->
            "$OPENCODE_MANAGED_PROVIDER_ID/${normalized.substringAfter('/')}"
        '/' in normalized -> normalized
        else -> "$OPENCODE_MANAGED_PROVIDER_ID/$normalized"
    }
}

private fun apiModelId(value: String): String = value.substringAfter('/', value)

private fun providerModelId(providerId: String, value: String): String {
    val normalized = RemoteCodexSettings.normalizeDefaultModel(value)
    return normalized.takeIf(String::isNotBlank)?.let { id ->
        if (id.startsWith("$providerId/")) id else "$providerId/$id"
    }.orEmpty()
}

/**
 * OpenCode adapter. Its REST/SSE translation runs in the managed remote bridge, while the
 * hardened SSH JSONL, SFTP, cache, cancellation, and event plumbing are shared with Codex.
 */
class OpenCodeAgentClient(
    scope: CoroutineScope,
    private val bridgeSource: () -> String,
    private val delegate: CodexAppServerClient = CodexAppServerClient(scope),
    private val bootstrapTransport: SshCodexTransport = SshCodexTransport(),
) : RemoteAgentClient by delegate {
    override val kind: AgentKind = AgentKind.OpenCode
    override val capabilities: AgentCapabilities = AgentCapabilities.OpenCode
    @Volatile
    private var settingsProviderId: String = OPENCODE_MANAGED_PROVIDER_ID
    /** Models added through the app are stable for one bridge generation. */
    private val ensuredCustomModels = ConcurrentHashMap<String, Long>()

    override suspend fun inspectRuntime(profile: ServerProfile): AgentRuntimeInspection {
        val host = delegate.inspectRuntime(profile)
        if (host.installationProblem != null) return host.copy(compatibleCommand = null)
        val values = bootstrapTransport.runBootstrapScript(
            profile = profile,
            script = OpenCodeBootstrap.probeScript,
            timeoutMs = PROBE_TIMEOUT_MS,
            operationName = "检测 OpenCode 运行时",
        ).let(OpenCodeBootstrap::parseProbe)
        val version = values["VERSION"]?.takeIf(String::isNotBlank)
        val bridge = values["BRIDGE"]?.takeIf(String::isNotBlank)
        val bridgeMatches = values["BRIDGE_SHA256"] == OpenCodeBootstrap.bridgeSha256(bridgeSource())
        val compatible = version?.contains(BuildConfig.PINNED_OPENCODE_VERSION) == true &&
            bridge != null && bridgeMatches
        return AgentRuntimeInspection(
            os = host.os,
            architecture = host.architecture,
            home = host.home,
            detectedVersion = version,
            compatibleCommand = bridge?.takeIf { compatible }?.let(::quoteShellArgument),
            installationProblem = host.installationProblem,
        )
    }

    override suspend fun installRuntime(
        profile: ServerProfile,
        onProgress: (RemoteInstallProgress) -> Unit,
    ) {
        bootstrapTransport.installNodeRuntimeDetailed(
            profile = profile,
            nodeVersion = BuildConfig.PINNED_NODE_VERSION,
        ) { progress ->
            onProgress(
                progress.copy(
                    percent = (progress.percent * SHARED_RUNTIME_PERCENT / 100)
                        .coerceIn(0, SHARED_RUNTIME_PERCENT),
                    message = "准备 OpenCode 运行时 · ${progress.message}",
                ),
            )
        }
        bootstrapTransport.runBootstrapScript(
            profile = profile,
            script = OpenCodeBootstrap.installScript(
                openCodeVersion = BuildConfig.PINNED_OPENCODE_VERSION,
                proxyUrl = profile.proxyUrl,
                bridgeSource = bridgeSource(),
            ),
            timeoutMs = INSTALL_TIMEOUT_MS,
            operationName = "安装 OpenCode",
        ) { line ->
            parseProgress(line)?.let(onProgress)
        }
    }

    override suspend fun uninstallRuntime(profile: ServerProfile) {
        ensuredCustomModels.clear()
        delegate.disconnect()
        bootstrapTransport.runBootstrapScript(
            profile = profile,
            script = OpenCodeBootstrap.uninstallScript,
            timeoutMs = UNINSTALL_TIMEOUT_MS,
            operationName = "卸载 OpenCode",
        )
        if (profile.agentMode == AgentMode.OpenCode) {
            delegate.uninstallRuntime(profile)
        }
    }

    override suspend fun connect(profile: ServerProfile): String {
        ensuredCustomModels.clear()
        val command = buildString {
            append(MANAGED_OPENCODE_BRIDGE_COMMAND)
            profile.workspace.takeIf(String::isNotBlank)?.let { workspace ->
                append(" --directory ").append(quoteShellArgument(workspace))
            }
        }
        return delegate.connect(profile.copy(remoteCommand = command))
    }

    override suspend fun disconnect() {
        ensuredCustomModels.clear()
        delegate.disconnect()
    }

    override suspend fun readGlobalSettings(profile: ServerProfile): AgentGlobalSettings {
        val result = delegate.requestAdapterExtension(
            method = "agent/settings/read",
            timeoutMs = SETTINGS_TIMEOUT_MS,
        ) as? JsonObject ?: error("OpenCode 设置返回格式错误")
        settingsProviderId = result.string("modelProvider").ifBlank { OPENCODE_MANAGED_PROVIDER_ID }
        return AgentGlobalSettings(
            baseUrl = result.string("baseUrl"),
            model = result.string("model"),
            reasoningEffort = result.string("reasoningEffort"),
            modelProvider = settingsProviderId,
            hasStoredAuthentication = (result["hasStoredAuthentication"] as? JsonPrimitive)
                ?.booleanOrNull == true,
            apiKey = result.string("apiKey"),
            proxyUrl = result.string("proxyUrl"),
        )
    }

    override suspend fun writeGlobalSettings(
        profile: ServerProfile,
        baseUrl: String,
        apiKey: String,
        proxyUrl: String,
        defaultModel: String,
        defaultReasoningEffort: String,
        preserveCurrentProvider: Boolean,
    ) {
        val normalizedBaseUrl = RemoteCodexSettings.validateBaseUrl(baseUrl)
        val normalizedProxy = RemoteBootstrap.validateProxyUrl(proxyUrl)
        val normalizedApiKey = apiKey.trim().also { key ->
            require(key.none { it.isWhitespace() || it.code !in 0x20..0x7e }) {
                "API 密钥不能包含空格、换行或控制字符"
            }
        }
        val normalizedDefaultModel = normalizeOpenCodeModelId(defaultModel)
        val normalizedDefaultReasoningEffort =
            RemoteCodexSettings.normalizeDefaultReasoningEffort(defaultReasoningEffort)
        val effectivePreserveCurrentProvider = preserveCurrentProvider &&
            (normalizedDefaultModel.isBlank() || normalizedDefaultModel.startsWith("$settingsProviderId/"))
        val result = delegate.requestAdapterExtension(
            method = "agent/settings/write",
            params = buildJsonObject {
                put("baseUrl", normalizedBaseUrl)
                put("apiKey", normalizedApiKey)
                put("proxyUrl", normalizedProxy)
                put("defaultModel", normalizedDefaultModel)
                put("defaultReasoningEffort", normalizedDefaultReasoningEffort)
                put("preserveCurrentProvider", effectivePreserveCurrentProvider)
                put("customModels", buildJsonArray {
                    profile.modelSettings(AgentKind.OpenCode).customModels.forEach { definition ->
                        add(buildJsonObject {
                            put("modelId", definition.modelId)
                            put("displayName", definition.displayName)
                            put("contextWindowTokens", definition.contextWindowTokens)
                            put("maxOutputTokens", definition.maxOutputTokens)
                        })
                    }
                })
            },
            timeoutMs = SETTINGS_TIMEOUT_MS,
        ) as? JsonObject
        settingsProviderId = result?.string("modelProvider")
            ?.ifBlank { OPENCODE_MANAGED_PROVIDER_ID }
            ?: OPENCODE_MANAGED_PROVIDER_ID
        ensuredCustomModels.clear()
    }

    override suspend fun testGlobalSettings(
        profile: ServerProfile,
        baseUrl: String,
        apiKey: String,
        proxyUrl: String,
        testModel: String,
    ): AgentConnectionTestResult = bootstrapTransport.testCodexGlobalSettings(
        profile = profile,
        baseUrl = baseUrl,
        apiKey = apiKey,
        proxyUrl = proxyUrl,
        testModel = apiModelId(testModel),
    )

    override suspend fun fetchApiModels(
        profile: ServerProfile,
        baseUrl: String,
        apiKey: String,
        proxyUrl: String,
    ): List<ApiModelOption> = bootstrapTransport.fetchApiModels(
        profile = profile,
        baseUrl = baseUrl,
        apiKey = apiKey,
        proxyUrl = proxyUrl,
    ).map { option ->
        option.copy(modelId = providerModelId(settingsProviderId, option.modelId))
    }

    override suspend fun ensureCustomModel(
        profile: ServerProfile,
        definition: CustomModelDefinition,
    ) {
        val generation = delegate.currentGeneration() ?: return
        val key = customModelKey(definition)
        if (ensuredCustomModels[key] == generation) return
        syncCustomModels(profile, listOf(definition), emptyList())
    }

    override suspend fun syncCustomModels(
        profile: ServerProfile,
        definitions: List<CustomModelDefinition>,
        removedModelIds: List<String>,
    ) {
        val generation = delegate.currentGeneration() ?: return
        val normalizedDefinitions = definitions.map { definition ->
            definition.copy(modelId = normalizeOpenCodeModelId(definition.modelId))
        }.distinctBy { customModelKey(it) }
        val normalizedRemovedIds = removedModelIds.asSequence()
            .map(::normalizeOpenCodeModelId)
            .filter(String::isNotBlank)
            .distinct()
            .toList()
        val pendingDefinitions = normalizedDefinitions.filter { definition ->
            ensuredCustomModels[customModelKey(definition)] != generation
        }
        if (pendingDefinitions.isEmpty() && normalizedRemovedIds.isEmpty()) return
        delegate.requestAdapterExtension(
            method = "agent/models/sync",
            params = buildJsonObject {
                put("models", buildJsonArray {
                    pendingDefinitions.forEach { definition ->
                        add(buildJsonObject {
                            put("modelId", definition.modelId)
                            put("displayName", definition.displayName)
                            put("contextWindowTokens", definition.contextWindowTokens)
                            put("maxOutputTokens", definition.maxOutputTokens)
                        })
                    }
                })
                put("removeModelIds", buildJsonArray {
                    normalizedRemovedIds.forEach(::add)
                })
            },
            timeoutMs = SETTINGS_TIMEOUT_MS,
        )
        if (delegate.isGenerationActive(generation)) {
            normalizedRemovedIds.forEach(::removeCachedModel)
            pendingDefinitions.forEach { definition ->
                removeCachedModel(definition.modelId)
                ensuredCustomModels[customModelKey(definition)] = generation
            }
        }
    }

    override fun close() {
        ensuredCustomModels.clear()
        delegate.close()
        bootstrapTransport.close()
    }

    companion object {
        const val MANAGED_OPENCODE_BRIDGE_COMMAND = "~/.local/bin/codex-remote-opencode-bridge"
        private const val SHARED_RUNTIME_PERCENT = 68
        private const val PROBE_TIMEOUT_MS = 30_000L
        private const val INSTALL_TIMEOUT_MS = 20 * 60_000L
        private const val UNINSTALL_TIMEOUT_MS = 60_000L
        private const val SETTINGS_TIMEOUT_MS = 30_000L

        private fun customModelKey(definition: CustomModelDefinition): String = listOf(
            normalizeOpenCodeModelId(definition.modelId),
            definition.displayName.trim(),
            definition.contextWindowTokens.toString(),
            definition.maxOutputTokens.toString(),
        ).joinToString("\u0000")

        private fun modelIdFromCustomModelKey(key: String): String = key.substringBefore('\u0000')

        private fun parseProgress(line: String): RemoteInstallProgress? {
            if (!line.startsWith("::progress::")) return null
            val fields = line.removePrefix("::progress::").split('|', limit = 4)
            if (fields.size < 3) return null
            return RemoteInstallProgress(
                percent = fields[0].toIntOrNull()?.coerceIn(0, 100) ?: 0,
                downloadPercent = fields[1].toIntOrNull()?.coerceIn(0, 100),
                message = fields[2],
                detail = fields.getOrElse(3) { "" },
            )
        }
    }

    private fun removeCachedModel(modelId: String) {
        ensuredCustomModels.keys.removeIf { key -> modelIdFromCustomModelKey(key) == modelId }
    }
}

File: flutter_app/lib/src/agent/opencode_agent_client.dart
import 'dart:async';

import '../domain/model_catalog.dart';
import '../domain/models.dart';
import '../ssh/ssh_server_client.dart';
import 'codex_agent_client.dart';
import 'codex_global_settings.dart';
import 'open_code_bootstrap.dart';
import 'open_code_bridge_asset.dart';
import 'remote_agent_client.dart';
import 'remote_bootstrap.dart';

typedef OpenCodeBridgeLoader = Future<String> Function();

const Duration _openCodeSettingsTimeout = Duration(seconds: 30);
const Duration _openCodeInstallTimeout = Duration(minutes: 30);

/// OpenCode adapter backed by the app-owned JSONL compatibility bridge.
///
/// The bridge deliberately mirrors the Codex app-server protocol, so thread,
/// turn, approval, notification, and generation handling stay in
/// [CodexAgentClient]. Only OpenCode-owned runtime and configuration behavior
/// lives here.
class OpenCodeAgentClient extends CodexAgentClient
    implements RemoteAgentCustomModelClient {
  OpenCodeAgentClient({
    super.clientVersion,
    super.requestTimeout,
    super.threadRequestTimeout,
    super.maxLineChars,
    super.sessionOpener,
    super.dedicatedHostFactory,
    OpenCodeBridgeLoader? bridgeLoader,
  }) : _bridgeLoader = bridgeLoader ?? OpenCodeBridgeAsset.load;

  final OpenCodeBridgeLoader _bridgeLoader;
  final Map<String, String> _ensuredCustomModels = <String, String>{};

  String _settingsProviderId = openCodeManagedProviderId;
  int _connectionEpoch = 0;

  @override
  AgentKind get kind => AgentKind.openCode;

  @override
  AgentCapabilities get capabilities => AgentCapabilities.openCode;

  @override
  Future<AgentRuntimeInspection> inspectRuntime(
    ServerProfile profile,
    RemoteServerClient host,
  ) async {
    final scriptHost = host is RemoteServerScriptClient
        ? host as RemoteServerScriptClient
        : throw UnsupportedError('当前 SSH 客户端不支持安全执行探测脚本');
    final bridgeSource = await _bridgeLoader();
    final output = await scriptHost.runShellScript(
      OpenCodeBootstrap.combinedProbeScript,
      timeout: const Duration(seconds: 30),
      maxOutputBytes: 64 * 1024,
    );
    return OpenCodeBootstrap.inspect(output, bridgeSource: bridgeSource);
  }

  @override
  Future<void> installRuntime(
    ServerProfile profile,
    RemoteServerClient host, {
    required void Function(RemoteInstallProgress progress) onProgress,
  }) async {
    final scriptHost = host is RemoteServerStreamingScriptClient
        ? host as RemoteServerStreamingScriptClient
        : throw UnsupportedError('当前 SSH 客户端不支持流式执行安装脚本');
    final bridgeSource = await _bridgeLoader();

    await scriptHost.runStreamingShellScript(
      OpenCodeBootstrap.installNodeRuntimeScript(proxyUrl: profile.proxyUrl),
      command: remoteInstallCommand,
      timeout: _openCodeInstallTimeout,
      maxOutputBytes: 8 * 1024 * 1024,
      onStdoutLine: (line) {
        final progress = parseRemoteInstallProgressLine(line);
        if (progress == null) return;
        onProgress(
          RemoteInstallProgress(
            percent: (progress.percent * openCodeSharedRuntimePercent ~/ 100)
                .clamp(0, openCodeSharedRuntimePercent),
            message: '准备 OpenCode 运行时 · ${progress.message}',
            detail: progress.detail,
            downloadPercent: progress.downloadPercent,
          ),
        );
      },
    );

    await scriptHost.runStreamingShellScript(
      OpenCodeBootstrap.installScript(
        proxyUrl: profile.proxyUrl,
        bridgeSource: bridgeSource,
      ),
      command: remoteInstallCommand,
      timeout: _openCodeInstallTimeout,
      maxOutputBytes: 8 * 1024 * 1024,
      onStdoutLine: (line) {
        final progress = parseRemoteInstallProgressLine(line);
        if (progress != null) onProgress(progress);
      },
    );
  }

  @override
  Future<void> uninstallRuntime(
    ServerProfile profile,
    RemoteServerClient host,
  ) async {
    await disconnect();
    final scriptHost = host is RemoteServerScriptClient
        ? host as RemoteServerScriptClient
        : throw UnsupportedError('当前 SSH 客户端不支持安全执行卸载脚本');
    await scriptHost.runShellScript(
      OpenCodeBootstrap.uninstallScript,
      timeout: const Duration(minutes: 1),
      maxOutputBytes: 64 * 1024,
    );
  }

  @override
  Future<void> connect(ServerProfile profile, RemoteServerClient host) async {
    _invalidateCustomModelCache();
    final command = buildOpenCodeBridgeCommand(profile.workspace);
    try {
      await super.connect(
        profile.copyWith(workspace: '', remoteCommand: command),
        host,
      );
    } catch (_) {
      _invalidateCustomModelCache();
      rethrow;
    }
  }

  @override
  Future<void> disconnect() async {
    _invalidateCustomModelCache();
    await super.disconnect();
  }

  @override
  Future<AgentGlobalSettings> readGlobalSettings(ServerProfile profile) async {
    final result = await requestAdapterExtension(
      'agent/settings/read',
      timeout: _openCodeSettingsTimeout,
    );
    final settings = parseOpenCodeGlobalSettings(result);
    _settingsProviderId = settings.modelProvider.trim().isEmpty
        ? openCodeManagedProviderId
        : settings.modelProvider.trim();
    return settings;
  }

  @override
  Future<void> writeGlobalSettings(
    ServerProfile profile, {
    required String baseUrl,
    required String apiKey,
    required String proxyUrl,
    required String defaultModel,
    required String defaultReasoningEffort,
    required bool preserveCurrentProvider,
  }) async {
    final normalizedBaseUrl = normalizeCodexBaseUrl(baseUrl);
    final normalizedApiKey = normalizeCodexApiKey(apiKey);
    final normalizedProxyUrl = normalizeCodexProxyUrl(proxyUrl);
    final normalizedDefaultModel = normalizeOpenCodeModelId(defaultModel);
    final normalizedEffort = normalizeCodexReasoningEffort(
      defaultReasoningEffort,
    );
    final effectivePreserveCurrentProvider =
        preserveCurrentProvider &&
        (normalizedDefaultModel.isEmpty ||
            normalizedDefaultModel.startsWith('$_settingsProviderId/'));
    final definitions = profile
        .modelSettings(AgentKind.openCode)
        .customModels
        .map(
          (definition) => normalizeCustomModelDefinitionForAgent(
            AgentKind.openCode,
            definition,
          ),
        )
        .toList(growable: false);

    final result = await requestAdapterExtension(
      'agent/settings/write',
      params: <String, Object?>{
        'baseUrl': normalizedBaseUrl,
        'apiKey': normalizedApiKey,
        'proxyUrl': normalizedProxyUrl,
        'defaultModel': normalizedDefaultModel,
        'defaultReasoningEffort': normalizedEffort,
        'preserveCurrentProvider': effectivePreserveCurrentProvider,
        'customModels': definitions.map(openCodeCustomModelParams).toList(),
      },
      timeout: _openCodeSettingsTimeout,
    );
    _settingsProviderId = openCodeSettingsProviderId(result);
    _invalidateCustomModelCache();
  }

  @override
  Future<AgentConnectionTestResult> testGlobalSettings(
    ServerProfile profile, {
    required String baseUrl,
    required String apiKey,
    required String proxyUrl,
    required String testModel,
    ModelApiProtocol? apiProtocol,
  }) {
    final definitions = profile.modelSettings(AgentKind.openCode).customModels;
    final resolvedProtocol =
        apiProtocol ?? resolveOpenCodeModelApiProtocol(testModel, definitions);
    return super.testGlobalSettings(
      profile,
      baseUrl: baseUrl,
      apiKey: apiKey,
      proxyUrl: proxyUrl,
      testModel: openCodeApiModelId(testModel),
      apiProtocol: resolvedProtocol,
    );
  }

  @override
  Future<List<ApiModelOption>> fetchApiModels(
    ServerProfile profile, {
    required String baseUrl,
    required String apiKey,
    required String proxyUrl,
  }) async {
    final options = await super.fetchApiModels(
      profile,
      baseUrl: baseUrl,
      apiKey: apiKey,
      proxyUrl: proxyUrl,
    );
    return List<ApiModelOption>.unmodifiable(
      options.map(
        (option) => option.copyWith(
          modelId: openCodeProviderModelId(_settingsProviderId, option.modelId),
        ),
      ),
    );
  }

  @override
  Future<void> ensureCustomModel(
    ServerProfile profile,
    CustomModelDefinition definition,
  ) async {
    final normalized = normalizeCustomModelDefinitionForAgent(
      AgentKind.openCode,
      definition,
    );
    if (_ensuredCustomModels.containsKey(
      openCodeCustomModelCacheKey(normalized),
    )) {
      return;
    }
    await syncCustomModels(
      profile,
      definitions: <CustomModelDefinition>[normalized],
      removedModelIds: const <String>[],
    );
  }

  @override
  Future<void> syncCustomModels(
    ServerProfile profile, {
    required List<CustomModelDefinition> definitions,
    required List<String> removedModelIds,
  }) async {
    if (!isConnected) return;
    final normalizedDefinitions = <CustomModelDefinition>[];
    final definitionKeys = <String>{};
    for (final definition in definitions) {
      final normalized = normalizeCustomModelDefinitionForAgent(
        AgentKind.openCode,
        definition,
      );
      if (definitionKeys.add(openCodeCustomModelCacheKey(normalized))) {
        normalizedDefinitions.add(normalized);
      }
    }
    final normalizedRemovedIds = <String>[];
    final removalIds = <String>{};
    for (final modelId in removedModelIds) {
      final normalized = normalizeOpenCodeModelId(modelId);
      if (normalized.isNotEmpty && removalIds.add(normalized)) {
        normalizedRemovedIds.add(normalized);
      }
    }
    final pendingDefinitions = normalizedDefinitions
        .where(
          (definition) => !_ensuredCustomModels.containsKey(
            openCodeCustomModelCacheKey(definition),
          ),
        )
        .toList(growable: false);
    if (pendingDefinitions.isEmpty && normalizedRemovedIds.isEmpty) return;

    final epoch = _connectionEpoch;
    await requestAdapterExtension(
      'agent/models/sync',
      params: <String, Object?>{
        'models': pendingDefinitions.map(openCodeCustomModelParams).toList(),
        'removeModelIds': normalizedRemovedIds,
      },
      timeout: _openCodeSettingsTimeout,
    );
    if (!isConnected || epoch != _connectionEpoch) return;

    for (final modelId in normalizedRemovedIds) {
      _removeCachedModel(modelId);
    }
    for (final definition in pendingDefinitions) {
      _removeCachedModel(definition.modelId);
      _ensuredCustomModels[openCodeCustomModelCacheKey(definition)] =
          definition.modelId;
    }
  }

  @override
  void close() {
    _invalidateCustomModelCache();
    super.close();
  }

  void _removeCachedModel(String modelId) {
    _ensuredCustomModels.removeWhere((_, cachedId) => cachedId == modelId);
  }

  void _invalidateCustomModelCache() {
    _ensuredCustomModels.clear();
    _connectionEpoch += 1;
  }
}

String buildOpenCodeBridgeCommand(String workspace) {
  final directory = workspace.trim();
  return directory.isEmpty
      ? managedOpenCodeBridgeCommand
      : '$managedOpenCodeBridgeCommand --directory ${shellQuote(directory)}';
}

AgentGlobalSettings parseOpenCodeGlobalSettings(Object? value) {
  final result = _stringKeyedMap(value);
  final provider = _mapString(result, 'modelProvider');
  return AgentGlobalSettings(
    baseUrl: _mapString(result, 'baseUrl'),
    model: _mapString(result, 'model'),
    reasoningEffort: _mapString(result, 'reasoningEffort'),
    modelProvider: provider.isEmpty ? openCodeManagedProviderId : provider,
    hasStoredAuthentication: result['hasStoredAuthentication'] == true,
    apiKey: _mapString(result, 'apiKey'),
    proxyUrl: _mapString(result, 'proxyUrl'),
  );
}

String openCodeSettingsProviderId(Object? value) {
  if (value is! Map) return openCodeManagedProviderId;
  final provider = _mapString(_stringKeyedMap(value), 'modelProvider');
  return provider.isEmpty ? openCodeManagedProviderId : provider;
}

Map<String, Object?> openCodeCustomModelParams(
  CustomModelDefinition definition,
) => <String, Object?>{
  'modelId': definition.modelId,
  'displayName': definition.displayName,
  'contextWindowTokens': definition.contextWindowTokens,
  'maxOutputTokens': definition.maxOutputTokens,
  'apiProtocol': switch (definition.apiProtocol) {
    ModelApiProtocol.chatCompletions => 'chat_completions',
    ModelApiProtocol.responses => 'responses',
  },
};

String openCodeCustomModelCacheKey(CustomModelDefinition definition) =>
    <String>[
      normalizeOpenCodeModelId(definition.modelId),
      definition.displayName.trim(),
      definition.contextWindowTokens.toString(),
      definition.maxOutputTokens.toString(),
      switch (definition.apiProtocol) {
        ModelApiProtocol.chatCompletions => 'chat_completions',
        ModelApiProtocol.responses => 'responses',
      },
    ].join('\u0000');

ModelApiProtocol resolveOpenCodeModelApiProtocol(
  String modelId,
  List<CustomModelDefinition> definitions,
) {
  final normalizedModelId = normalizeOpenCodeModelId(modelId);
  for (final definition in definitions) {
    if (normalizeOpenCodeModelId(definition.modelId) == normalizedModelId) {
      return definition.apiProtocol;
    }
  }
  return ModelApiProtocol.chatCompletions;
}

String openCodeApiModelId(String value) {
  final normalized = normalizeCodexModel(value, '测试模型');
  final separator = normalized.indexOf('/');
  return separator < 0 ? normalized : normalized.substring(separator + 1);
}

String openCodeProviderModelId(String providerId, String value) {
  final provider = providerId.trim().isEmpty
      ? openCodeManagedProviderId
      : providerId.trim();
  final normalized = normalizeCodexModel(value, '模型');
  if (normalized.isEmpty || normalized.startsWith('$provider/')) {
    return normalized;
  }
  return '$provider/$normalized';
}

Map<String, Object?> _stringKeyedMap(Object? value) {
  if (value is! Map) throw StateError('OpenCode 设置返回格式错误');
  return <String, Object?>{
    for (final entry in value.entries)
      if (entry.key is String) entry.key as String: entry.value,
  };
}

String _mapString(Map<String, Object?> value, String key) {
  final item = value[key];
  return item is String ? item.trim() : '';
}
